src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -51,46 +52,121 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
-
-                        fields = ImmutableArray.Create(
-                                    new TupleFieldSymbol(elementNames.IsDefault ? "Item1" : elementNames[0],
-                                        this,
-                                        elementTypes[0],
-                                        underlyingField1?.AsMember(underlyingType)),
-                                    new TupleFieldSymbol(elementNames.IsDefault ? "Item2" : elementNames[1],
-                                        this,
-                                        elementTypes[1],
-                                        underlyingField2?.AsMember(underlyingType))
-                                );
-
-                        return underlyingType;
-                    }
+                var fieldSymbols = ArrayBuilder<TupleFieldSymbol>.GetInstance(elementTypes.Length);
 
-                default:
-                    {
-                        // TODO: VS if this eventually still stays reachable, need to make some error type symbol
-                        var tupleType = binder.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_ValueTuple_T1_T2, diagnostics, syntax);
-                        underlyingType = tupleType.Construct(elementTypes);
-                        break;
-                    }
+                for (int i = 0; i < elementTypes.Length; i++)
+                {
+                    var wellKnownTypeMember = EnumTypeMember(elementTypes.Length, i);
+                    var underlyingField = (FieldSymbol)Binder.GetWellKnownTypeMember(binder.Compilation, wellKnownTypeMember, diagnostics, syntax: syntax);
+                    var field = new TupleFieldSymbol(elementNames.IsDefaultOrEmpty ? UnderlyingMemberName(i) : elementNames[i],
+                                               this,
+                                               elementTypes[i],
+                                               underlyingField?.AsMember(underlyingType));
+                    fieldSymbols.Add(field);
+                }
+
+                fields = ImmutableArray.Create(fieldSymbols.ToImmutableAndFree(), 0, elementTypes.Length);
+            }
+            else
+            {
+                fields = ImmutableArray<TupleFieldSymbol>.Empty;
 
+                // PROTOTYPE: VS if this eventually still stays reachable, need to make some error type symbol
             }
 
             return underlyingType;
         }
 
+        /// <summary>
+        /// Find the well-known ValueTuple type of a given arity.
+        /// For example, for arity=2:
+        /// returns WellKnownType.System_Runtime_CompilerServices_ValueTuple_T1_T2
+        /// </summary>
+        private static WellKnownType EnumType(int arity)
+        {
+            return tupleTypes[arity - 1];
```

Comment:
@AlekseyTs Added a test to cover the matrix and did find a bug/typo. Thanks!
I didn't test it directly though, as I ran into accessibility problems (internal). I'm sure there is a trick to workaround, but I don't know it yet...